src/metrics/handle.rs

```diff
@@ -0,0 +1,257 @@
+use metrics::GaugeValue;
+use std::sync::atomic::{AtomicU32, AtomicU64, Ordering};
+
+#[derive(Debug)]
+pub enum Handle {
+    Gauge(Gauge),
+    Counter(Counter),
+    Histogram(Histogram),
+}
+
+impl Handle {
+    pub(crate) fn counter() -> Self {
+        Handle::Counter(Counter::new())
+    }
+
+    pub(crate) fn increment_counter(&mut self, value: u64) {
+        match self {
+            Handle::Counter(counter) => counter.record(value),
+            _ => unreachable!(),
+        }
+    }
+
+    pub(crate) fn gauge() -> Self {
+        Handle::Gauge(Gauge::new())
+    }
+
+    pub(crate) fn update_gauge(&mut self, value: GaugeValue) {
+        match self {
+            Handle::Gauge(gauge) => gauge.record(value),
+            _ => unreachable!(),
+        }
+    }
+
+    pub(crate) fn histogram() -> Self {
+        Handle::Histogram(Histogram::new())
+    }
+
+    pub(crate) fn record_histogram(&mut self, value: f64) {
+        match self {
+            Handle::Histogram(h) => h.record(value),
+            _ => unreachable!(),
+        };
+    }
+}
+
+#[derive(Debug)]
+pub struct Histogram {
+    pub buckets: Box<[(f64, AtomicU32); 22]>,
+    pub count: AtomicU32,
+    pub sum: AtomicU64,
+}
+
+impl Histogram {
+    pub(crate) fn new() -> Self {
+        // Box to avoid having this large array inline to the structure, blowing
+        // out cache coherence.
+        //
+        // The sequence here is based on powers of two. Other sequences are more
+        // suitable for different distributions but since our present use case
+        // is mostly non-negative and measures smallish latencies we cluster
+        // around but never quite get to zero with an increasingly coarse
+        // long-tail.
+        let buckets = Box::new([
```

Comment:
Possibly relevant: https://github.com/timberio/vector/issues/6546

I am a little concerned about a one-size-fits-all set of buckets. Do you think this is something we might be able to configure when creating new histograms (i.e. pass through the `new()` here)?

Reply:
Overall 👍 to the strategy though. This should bound the memory use.

Reply:
I don't love the one-size notion either. The only code that will know its distribution is the caller, is my argument. I _think_ it's possible to work ourselves into a good situation here, but it's something I'd like to tackle in a follow-up PR to avoid bloating this PR. 

My basic idea comes down to an observation that we spend _a lot_ of runtime dropping temporary `Key` instances. Our total number of keys is relatively low but we need to make a new one of these each time we do a recording. Seems like if we could adjust call-sites to not use the macro but instead request a handle that they then hold and record though we could address that churn _and_ allow the call-site to set their own bounds.

Reply:
Hmm, this reminds me that I need to check and make sure that the prometheus sink picks up these buckets properly.

Reply:
Agreed, we can follow up on letting callers configure.

In addition to prometheus, it might be worth checking our other metrics sinks too.

Reply:
Excellent. So long as we pass through an `AggregatedHistogram` -- which we are now -- it looks like everything downstream inherits the buckets from that type.

Reply:
Ironically, dedicated handles that mapped to a specific key (name + labels) is what metrics 0.12 did and was actually about as performant as possible. 😂 

Interested to hear more about this, though i.e. are the keys actually static? etc.

Reply:
> Interested to hear more about this, though i.e. are the keys actually static? etc.

There's enough spots in our code base where the key and labels are static in a hot path and there's a place for me to stash a reference to the handle in a struct that it'd be worth giving it a shot. That said, there are for sure spots in the code base where the key is fully dynamic, so it's not a blanket approach.